src/artm/core/cooccurrence_collector.cc

```diff
@@ -0,0 +1,1053 @@
+  // 2. For every first token id create an object Cell and for every second token
+  // that co-occurred with first write it's id, cooc_tf, cooc_df
+  // 3. Write the cell in output file and continue the cicle while there are
+  // first token ids in cooccurrence statistics holder
+  // Note that there can't be two cells stored in ram simultaniously
+  // 4. Save batch in vector of objects
+  std::shared_ptr<CooccurrenceBatch> batch(CreateNewCooccurrenceBatch());
+  OpenBatchOutputFile(batch);
+  for (auto iter = cooc_stat_holder.storage_.begin(); iter != cooc_stat_holder.storage_.end(); ++iter) {
+    batch->FormNewCell(iter);
+    batch->WriteCell();
+  }
+  CloseBatchOutputFile(batch);
+  vector_of_batches_.push_back(std::move(batch));
+}
+
+CooccurrenceBatch* CooccurrenceCollector::CreateNewCooccurrenceBatch() const {
+  return new CooccurrenceBatch(config_.target_folder());
+}
+
+void CooccurrenceCollector::OpenBatchOutputFile(std::shared_ptr<CooccurrenceBatch> batch) {
+  if (!batch->out_batch_.is_open()) {
+    assert(open_files_counter_ < max_num_of_open_files_);
+    ++open_files_counter_;
+    batch->out_batch_.open(batch->filename_, std::ios::out);
+  }
+}
+
+void CooccurrenceCollector::CloseBatchOutputFile(std::shared_ptr<CooccurrenceBatch> batch) {
+  if (batch->out_batch_.is_open()) {
+    --open_files_counter_;
+    batch->out_batch_.close();
+  }
+}
+
+unsigned CooccurrenceCollector::CooccurrenceBatchesQuantity() const {
+  return vector_of_batches_.size();
+}
+
+void CooccurrenceCollector::ReadAndMergeCooccurrenceBatches() {
+  // After that all the statistics has been gathered and saved in form of cooc batches on disk, it
+  // needs to be read and merged from cooc batches into one storage
+  // If number of cooc batches <= number of files than can be open simultaniously, then
+  // all the cooc batches are divided equally into n groups where n = number of cores.
+  // Else maximal number of open files is taken and divided into n groups.
+  // After that 1 thread has eliberated that can handle another portion of batches than aren't merged yet
+  // There are two stages of merging:
+  // 1. merging of files of one group (is done asynchroniously, without dropping of rare pairs of tokens)
+  // n files are written back in the same form (of cooccurrence batches)
+  // If n is too large to merge all the batches into some small number of batches
+  // this operation can be in cicle launched many times.
+  // 2. then that n files need to be read and merged again (with dropping of rare pairs of tokens)
+  // Merging of k files is implemented in KWayMerge function
+  // After the second stage the data is written in format of output file (not in format of cooc batches)
+  // If there would be a need to calculate ppmi or other values which depend on co-occurrences
+  // this data can be read back from output file.
+
+  // std::cout << "Step 2: merging batches" << std::endl;
+  const unsigned min_num_of_batches = 32;
```

Comment:
const in common.h

Reply:
Yeah, what's the sense in putting some variable in common.h if it's used only in this module? It breaks the sense of word 'common'